zebra-state/src/service/finalized_state/disk_format/upgrade.rs

```diff
@@ -272,180 +275,149 @@ impl DbFormatChange {
-
-            // Set up task for reading orchard note commitment trees
-            let orchard_compare_task = std::thread::spawn(move || {
-                while let Ok((height, tree)) = orchard_tree_rx.recv() {
-                    let tree = Some(tree);
-                    if prev_orchard_tree != tree {
-                        if let Err(error) = unique_orchard_tree_height_tx.send(height) {
-                            warn!(?error, "unexpected send error")
-                        }
+            // The Sapling note commitment trees are pruned if `last_height` has reached the height
+            // right after `initial_tip_height`. If `last_height` has not reached that height, they
+            // have been pruned in a previous upgrade, so the inclusive range between `last_height`
+            // and `initial_tip_height` must be empty. If it is not, there is an error, and the
+            // current upgrade did not succeed.
+            if last_height != initial_tip_height.next()
+                && db
+                    .sapling_tree_by_height_range(last_height..=initial_tip_height)
+                    .next()
+                    .is_some()
+            {
+                panic!("Zebra could not prune all Sapling trees.");
+            }
 
-                        prev_orchard_tree = tree;
-                    }
+            // Prune duplicate Orchard note commitment trees.
+            let mut last_tree = db.orchard_tree_by_height(&Height(0)).expect(
+                "The Orchard note commitment tree for the genesis block should be in the database.",
+            );
+            let mut last_height = Height(1);
+            // Run through all the trees in the finalized chain.
+            for (height, tree) in db.orchard_tree_by_height_range(Height(1)..=initial_tip_height) {
+                // Return early if there is a cancel signal.
+                if !matches!(cancel_receiver.try_recv(), Err(mpsc::TryRecvError::Empty)) {
+                    return;
                 }
-            });
-
-            // Set up task for deleting orchard note commitment trees
-            let db = upgrade_db;
-            let orchard_delete_task = std::thread::spawn(move || {
-                let mut delete_from = orchard_height.next();
-                while let Ok(delete_to) = unique_orchard_tree_height_rx.recv() {
-                    let num_entries = delete_to - delete_from;
-                    if num_entries > 0 {
+                // We delete duplicate trees in batches. A batch is a (possibly empty) series of
+                // identical trees excluding the first tree. We exclude the first tree so that we
+                // don't delete it. We get a batch if we encounter a tree that differs from the
+                // previous one, or if we're at the end of the chain. We get an empty batch if there
+                // are two consecutive differing trees.
+                if last_tree != tree || height == initial_tip_height {
```

Comment:
Same here

Reply:
This might be the cause of the panic I get on my testnet state:

```
The application panicked (crashed).
Message:  Zebra could not prune all Orchard trees.
Location: zebra-state/src/service/finalized_state/disk_format/upgrade.rs:409

Metadata:
version: 1.1.0+62.g961d45c
Zcash network: Testnet
running state version: 25.1.1
initial disk state version: 25.1.0
features: default,getblocktemplate_rpcs,metrics_exporter_prometheus,prometheus,release_max_level_info 
branch: state-dedup-fixes
git commit: 961d45c
```

You might want to turn this panic into an error log, because the detailed checks show which trees are still duplicates, but this panic doesn't.

Reply:
This edge case is addressed in [`4b579a6` (#7379)](https://github.com/ZcashFoundation/zebra/pull/7379/commits/4b579a60786751719110aee6fcfdde9262b02aef).